src/Infrastructure/Tests/Riganti.Utils.Infrastructure.EntityFramework.Tests/UnitOfWork/EntityFrameworkUnitOfWorkTests.cs

```diff
@@ -113,6 +184,74 @@ public async Task CommitAsync_CallCommitCoreOnlyIfHasOwnDbContext()
             unitOfWorkParentMock.Protected().Verify("CommitAsyncCore", Times.Once(), new CancellationToken());
         }
 
+        [Fact]
+        public async Task CommitAsync_ThrowIfChildCommitRequestedNotFulfilledByRoot()
+        {
+            Func<DbContext> dbContextFactory = () => new Mock<DbContext>().Object;
+
+            var unitOfWorkRegistryStub = new ThreadLocalUnitOfWorkRegistry();
+
+            var unitOfWorkProvider = new EntityFrameworkUnitOfWorkProvider(unitOfWorkRegistryStub, dbContextFactory);
+
+            await Assert.ThrowsAsync<InvalidOperationException>(async () =>
+            {
+                using (unitOfWorkProvider.Create(DbContextOptions.ReuseParentContext))
+                {
+                    using (var unitOfWorkChild = unitOfWorkProvider.Create(DbContextOptions.ReuseParentContext))
+                    {
+                        await unitOfWorkChild.CommitAsync();
+                    }
+                }
+            });
+
+            // test that unit of work provider keeps working after caught exception
+            using (unitOfWorkProvider.Create(DbContextOptions.ReuseParentContext))
+            {
+                using (unitOfWorkProvider.Create(DbContextOptions.ReuseParentContext))
+                {
+                }
+            }
+        }
+
+        [Fact]
+        public async Task CommitAsync_CorrectMultipleLayeredReuseParentBehavior()
+        {
+            Func<DbContext> dbContextFactory = () => new Mock<DbContext>().Object;
+
+            var unitOfWorkRegistryStub = new ThreadLocalUnitOfWorkRegistry();
```

Comment:
You should use `AsyncLocalUnitOfWorkRegistry` for async calls.

Reply:
ThreadLocalUnitOfWorkRegistry is used for consistency among tests.